python/cudf/cudf/core/frame.py

```diff
@@ -3581,64 +3582,148 @@ def _reindex(
+        Contains the values to be replaced.
+    value : numeric, str, list-like, or dict
+        Contains the values to replace `to_replace` with.
+    columns : dict
+        A column to dtype mapping representing dtype of columns.
+
+    Returns
+    -------
+    all_nan_columns : dict
+        A dict mapping of all columns if they contain all na values
+    to_replace_columns : dict
+        A dict mapping of all columns and the existing values that
+        have to be replaced.
+    values_columns : dict
+        A dict mapping of all columns and the corresponding values
+        to be replaced with.
+    """
+    to_replace_columns: Dict[Any, Any] = {}
+    values_columns: Dict[Any, Any] = {}
+    all_nan_columns: Dict[Any, Any] = {}
+
+    if is_scalar(to_replace) and is_scalar(value):
+        to_replace_columns = {col: [to_replace] for col in columns}
+        values_columns = {col: [value] for col in columns}
+    elif cudf.utils.dtypes.is_list_like(to_replace) or isinstance(
+        to_replace, cudf.core.column.ColumnBase
     ):
-        replacement = list(to_replace.values())
-        to_replace = list(to_replace.keys())
-    elif not is_scalar(to_replace):
-        if is_scalar(replacement):
-            all_nan = replacement is None
-            if all_nan:
-                replacement = [replacement] * len(to_replace)
-            # Do not broadcast numeric dtypes
-            elif pd.api.types.is_numeric_dtype(column.dtype):
-                if len(to_replace) > 0:
-                    replacement = [replacement]
-                else:
-                    # If to_replace is empty, replacement has to be empty.
-                    replacement = []
-            else:
-                replacement = utils.scalar_broadcast_to(
-                    replacement,
-                    (len(to_replace),),
-                    np.dtype(type(replacement)),
-                )
-        else:
-            # If both are non-scalar
-            if len(to_replace) != len(replacement):
+        if is_scalar(value):
+            to_replace_columns = {col: to_replace for col in columns}
+            values_columns = {
+                col: [value]
+                if pd.api.types.is_numeric_dtype(columns[col])
+                else [value] * len(to_replace)
+                for col in columns
+            }
+        elif cudf.utils.dtypes.is_list_like(value):
+            if len(to_replace) != len(value):
                 raise ValueError(
                     f"Replacement lists must be "
                     f"of same length."
-                    f"Expected {len(to_replace)}, got {len(replacement)}."
+                    f"Expected {len(to_replace)}, got {len(value)}."
                 )
-    else:
-        if not is_scalar(replacement):
+            else:
+                to_replace_columns = {col: to_replace for col in columns}
+                values_columns = {col: value for col in columns}
+        elif isinstance(value, cudf.Series):
```

Comment:
Do we need to special case `cudf.Series` here? We should avoid depending on `Series` whenever possible inside of `Frame` since `Series` inherits from `Frame`.

Reply:
yes, but we could use `is_column_like` where we have a check for either `cudf.Series` or `__cuda_array_interface__` & its shape.

Reply:
Do we need to do any index handling? We can't always use `__cuda_array_interface__` since it doesn't handle strings / lists / structs / etc.

If we don't need index handling could we just do `elif isinstance(value, Frame):`?

We also have bigger issues across Frame of using `Series` / `DataFrame` / `MultiIndex` where we can just continue for now and fix it later.

Reply:
> If we don't need index handling could we just do elif isinstance(value, Frame):?

We don't need index handling in this case, but wouldn't we be allowing `DataFrame`/ `MulitIndex` into this part of the code-flow instead of throwing an error ?

Reply:
We could check that `value` is a Frame and only has a single column.

Otherwise, should this function live somewhere else instead of in `frame.py`?

Reply:
How about moving `_get_replacement_values_for_columns` to `cudf.utils.utils`?(since we also have it documented clearly now).

Reply:
Seems more specialized than a general utility to me.

Reply:
Updated the code to use `Frame` based check for series.